### PR TITLE
216 td fix inverted reporter mode assignment

### LINKED ISSUE #216: [TD] Fix Inverted Reporter Mode Assignment
## Summary
Reporter selection inside `ReporterService` is inverted. Pipelines configured for `feedback_mode="ai"` instantiate `DefaultReporter`, while every other mode instantiates `AiReporter`. The service also lacks a `generate_feedback()` façade, so `FeedbackStep` crashes when it tries to call it.
## Scope
- Update `ReporterService` so `feedback_mode == "ai"` selects `AiReporter`, otherwise default to `DefaultReporter`.
- Add a `generate_feedback(grading_result, feedback_config)` method that parses config into `FeedbackPreferences` and delegates to the active reporter.
- Ensure `FeedbackStep` (and other callers) invoke the façade.
## Impacted Files
- autograder/services/report/reporter_service.py
- autograder/steps/feedback_step.py
## Risks
- Another logic inversion would silently break AI/default modes; cover both paths with tests.
- Reporter instantiation happens during pipeline build, so keep dependency injection intact.
## Acceptance Criteria
- Reporter mode selection matches config.
- `ReporterService.generate_feedback()` returns the reporter output.
- Unit tests cover default vs AI modes and the façade path.
## References
- docs/roadmaps/TECHNICAL_DEBT_ROADMAP.md (Item 1)

### LINKED ISSUE #217: [TD] Implement DefaultReporter Output
## Summary
`DefaultReporter.generate_report()` currently returns nothing, so pipelines in default feedback mode produce `None`. The `FeedbackPreferences` model and `ResultTreeFormatter` are already in place but unused.
## Scope
- Implement `DefaultReporter.generate_report()` to render readable text/markdown using the `ResultTree`, focus data, and `FeedbackPreferences`.
- Consume `DefaultReporterPreferences.category_headers` and `GeneralPreferences` visibility toggles when building the report.
- Add tests covering representative result trees to ensure formatting stability.
## Impacted Files
- autograder/services/report/default_reporter.py
- autograder/utils/formatters/result_tree_formatter.py (if helpers need adjustments)
## Risks
- Poor formatting or ignoring preference flags would make student feedback unusable; add snapshot/unit tests to guard output.
- Report generation should degrade gracefully when optional focus data is missing.
## Acceptance Criteria
- Default feedback mode returns a non-empty, structured report honoring preference settings.
- Unit tests cover baseline, focus-heavy, and preference-toggled scenarios.
## References
- docs/roadmaps/TECHNICAL_DEBT_ROADMAP.md (Item 2)

### LINKED ISSUE #219: [TD] Fix FeedbackStep Contract
## Summary
`FeedbackStep` calls `ReporterService.generate_feedback()` even though the service lacks that method, and it returns raw strings instead of `StepResult`. The step cannot currently complete without crashing.
## Scope
- After `ReporterService.generate_feedback()` exists (TD #216), update `FeedbackStep` to call it and wrap responses in `StepResult` with `StepName.FEEDBACK` and `StepStatus.SUCCESS`.
- Ensure failure paths return `StepStatus.FAIL` with useful error info.
## Impacted Files
- autograder/steps/feedback_step.py
- autograder/models/dataclass/step_result.py (if additional helpers are needed)
## Risks
- Misconstructed step results could break pipeline completion in `AutograderPipeline.finish_execution()`.
## Acceptance Criteria
- Feedback step returns a valid `StepResult` containing the reporter output.
- Tests cover success + failure flows.
## References
- docs/roadmaps/TECHNICAL_DEBT_ROADMAP.md (Item 4)

### PR DESCRIPTION
**Pull request by matheusmra**
This pull request introduces a comprehensive refactor and enhancement of the feedback/report generation system in the autograder, including improvements to the reporting logic, service delegation, and integration with pipeline steps. The changes also add robust unit and playroom tests to ensure reliability and correctness. The most important changes are grouped below:

### Reporting Logic Improvements

* Refactored `DefaultReporter.generate_report` to produce structured, customizable reports based on user preferences, including titles, summaries, categorized results, scores, and learning resources.
* Updated `AiReporter.generate_report` to accept new parameters (`focus`, `result_tree`, `preferences`) for future extensibility and compatibility with the new reporting interface.

### Service Delegation and Integration

* Refactored `ReporterService` to select the correct reporter based on `feedback_mode`, parse feedback config into preference objects, and delegate report generation to the active reporter. Added a new `generate_feedback` method for unified report creation.
* Updated `FeedbackStep.execute` to use the new `ReporterService.generate_feedback` method, properly handle success/failure, and utilize new factory methods for `StepResult`. [[1]](diffhunk://#diff-f566d82930bd6df0533e734f3f4574f011a0729da3e9475a13cccf7ff0433ecbL29-R52) [[2]](diffhunk://#diff-9ed56ad7762d7b19f13b53055b16e2aae24f15d04e23e27e13952b74c5d9002cR35-R44)

### Testing Enhancements

* Added playroom-style integration test for `DefaultReporter` to demonstrate report generation with mock data and preferences.
* Added unit tests for `DefaultReporter` covering report structure, passed test filtering, and graceful handling of empty input.
* Added unit tests for `ReporterService` to verify reporter selection, feedback delegation, and preference parsing.
* Added integration test for `FeedbackStep` to ensure correct execution and report content in the pipeline.

### Feedback Examples

Below are various configurations of report output based on feedback settings

<details>
<summary><b>Example 1: Full Detailed Report</b> (Summary, Categorized Tests, Score, and Resources)</summary>

# 🚀 Relatório de Desempenho do Aluno

---
## 📊 RESUMO
- 📋 **Template:** `projeto_final_python`
- 📊 **Tests:** 4 total  
  - ✅ 2 passed  
  - ❌ 2 failed

---
### 🧪 Soma de Inteiros ✅
- 📁 **Arquivo:** `main.py`
- 🟢 **Nota:** 100.0/100
- ⚙️ **Parâmetros:** `stdin=input_val`
- 💬 **Feedback:** Passou em todos os casos.

### 🧪 Subtração de Inteiros ❌
- 📁 **Arquivo:** `main.py`
- 🔴 **Nota:** 0.0/100
- ⚙️ **Parâmetros:** `stdin=input_val`
- 💬 **Erro:** Erro: 5 - 3 deveria ser 2, mas foi 8.

---
## 🏆 Nota Final

> **85.00 / 100**

---
## 📚 Recursos de Aprendizado

- 📘 [Tutorial de Controle de Fluxo](https://docs.python.org/3/tutorial/controlflow.html)

</details>

<details>
<summary><b>Example 2: Summary Only</b> (Passed tests hidden via <code>show_passed_tests=False</code>)</summary>

# Relatório de Avaliação

---
## 📊 RESUMO
- 📋 **Template:** `intro_python`
- 📊 **Tests:** 5 total  
  - ✅ 5 passed  
  - ❌ 0 failed

---
## 🏆 Nota Final

> **100.00 / 100**

</details>

<details>
<summary><b>Example 3: Minimalist</b> (Title and Score only)</summary>

# Relatório de Avaliação

---
## 🏆 Nota Final

> **100.00 / 100**

</details>

### CODE REVIEW

**Comment by matheusmra** [comment]
@ArthurCRodrigues ready for review :)

**Review by ArthurCRodrigues — CHANGES_REQUESTED** [review]
@matheusmra can you provide some feedback examples in the PR description?

**Comment by matheusmra** [comment]
> @matheusmra can you provide some feedback examples in the PR description?

Good morning @ArthurCRodrigues !
I just updated the pr description.

File: autograder/services/report/ai_reporter.py [hunk 1]
```diff
@@ -1,5 +1,8 @@
...
-class AiReporter:
-    def generate_report(self, result):
-        # Placeholder for AI report generation logic
-        return "AI-generated report based on analysis data."
\ No newline at end of file
+
+class AiReporter:
+    """Reporter that uses AI to generate feedback."""
+    def generate_report(self, _focus, _result_tree, _preferences=None):
+        """Generates an AI-based report based on analysis results."""
```
Comment (ArthurCRodrigues) [inline autograder/services/report/ai_reporter.py]: Can you make it throw a `NotImplemented` error instead of a mock?

**Review by ArthurCRodrigues — CHANGES_REQUESTED** [review]

File: autograder/services/report/ai_reporter.py (hunk 1 above)
Comment (matheusmra) [reply autograder/services/report/ai_reporter.py]: Done

**Review by ArthurCRodrigues — APPROVED** [review]
@matheusmra thanks! Just to let you know: i'm planning on creating a roadmap for the reporter service to have some major improvements, so these implementations will probably be refactored in a near future. 

I'll provide some more details soon, but nothing game changing.
